### ISSUE TITLE
Can you make it work with .NET 7.0 or .NET 8.0?

### ISSUE THREAD

**Issue by nsulikowski** [Excel-DNA/ExcelDna#538]
Or give me some hints to do it myself. I tried modifying the source code, but failed :-(
Thanks

**Comment by govert** [comment]
There is a hard limitation that different versions of the .NET core runtime cannot be loaded into a process. This means an instance of Excel can't load an add-in targeting .NET 6 and an add-in targeting .NET 7 at the same time.

For now I've decided to only support .NET 6 with Excel-DNA, so that different add-in can continue to work together without problems.

What are the main drivers for you to go towards .NET 7?

**Comment by nsulikowski** [comment]
All the libraries i wrap with ExcelDNA have been converted to NET 7
Can’t move back.
________________________________

**Comment by nsulikowski** [comment]
I figured out a way to go back to NET 6 for libraries. 
Thanks for your response. I think there should be an ExcelDna for .Net Framework, and another one for .NET 6, 7, etc. 
Thanks for your great work over the years.
N

**Comment by govert** [comment]
Just to be clear on the situation today. You can use the current version of Excel-DNA to make add-in that target the .NET Framework 4.6.2 and higher, and you can also make add-ins that target .NET 6. Such add-in can all run together in a single Excel process, mixing .NET Framework and .NET 6 add-ins together.

If we added support for .NET 7, then such an add-in could not run at the same time as a .NET 6 add-in, so depending on which add-in is loaded first, the other would fail to load. I'm trying to avoid having to deal with such problems by only support .NET 6 for now. Since .NET 6 is a long-term support version, that seems a reasonable solution for now. Over the coming years we would have to figure out what to do, but the concurrent loading constraint together with the yearly major .NET versions makes life hard for our use case.

**Comment by nsulikowski** [comment]
Yep, I get it. Thanks again.

**Comment by lilkui** [comment]
Native AOT is supported in .NET 7. Is it possible that Excel-DNA could support Native AOT deployment if targeting .NET 7?

**Comment by govert** [comment]
@lilkui Indeed, I think NativeAOT is a promising future direction for Excel-DNA, especially with the COM support being worked on for .NET 8. We're likely to do some work toward this next year - signing up as a [GitHub Sponsor](https://github.com/sponsors/Excel-DNA/) is one way to encourage work like this.

**Comment by CorsairRO** [comment]
Hello,

1) Thank you very much for all the effort that you put into that project.
2) I know the bitchyness of Excel and how it cannot load many .NET at the same time.
This is basically why Microsoft didnt move their VSTO out of .NET Framework, and following that on Microsoft it doesnt seem to be a solution soon or a willing for them to migrate it.
3) The alternative that Microsoft suggests there on their developer forums is just a bad joke, the Javascript new API.
4) I have played with this package and it seems again wonderful how 1 or a few people can make something to work nice while Microsoft with their huge teams cannot make their own things to work like they should
5) The package is exactly what i have been expected to be, and for that i have to again thank you very much.

6) However i dont understand why it should be your problem as package creator if i, as a user, am an idiot and want to target multiple addins on different frameworks.

7) Moving from .NET6 to .NET7 is after my tests a close to 20% improvement in perf in pretty much every area.
Not to mention the new revolution again that will be added to .NET8 this autumn.
8) All my projects are moved in a few days after the .NET is released even if they are very complex ones.
Excel COM Addins ( VSTO ) are the only ones that i couldnt move there.
9) So i really dont understand why a nice project like this one should be stuck to .NET6 because it might be consumed wrong by the users?

Sorry for the very long comment there,
Thank you again for supporting this project,
Have a nice day

**Comment by govert** [comment]
I appreciate the feedback and will try to canvas for some more opinions on what to do about .NET 7+.

My main concern is this. Some users of Excel-DNA create add-ins that are used beyond their organization, e.g. they ship their add-in as part of a product to outside customers, or the add-in (like the Excel-DNA IntelliSense extension) is used in various settings that might not carefully manage the .NET environment. If we support making add-ins with .NET 7+, any add-in targeting .NET 7+ will block all add-ins targeting .NET 6. And conversely, any add-in targeting .NET 6 will block all add-ins that target .NET 7+. This would mean the only viable runtime target for add-ins to be used in less controlled internal environments, or by outside users, would be .NET Framework 4.x.  So, by allowing .NET 7 to be used, we remove the .NET 6 option from many scenarios. To me it would be frustrating for all those projects to be stuck on .NET Framework for many years to come.

Over the long term there is one source of hope. Microsoft has put some effort in the pre-compiled NativeAOT option, with quite a bit of progress in .NET 8. It might be possible to engineer Excel-DNA to target NativeAOT, and in that way make a native code add-in based on the .NET 8+ runtime. These would not clash with any other add-ins, freeing us from the runtime restriction. This is not trivial from the Excel-DNA side (where reflection, runtime compilation and COM features are used extensively) so might take a while. Also, a NativeAOT add-in would require all the libraries used to also work with AOT, which might be a problem for a few years, since it's all really new. So that seems like one plan, even if it will probably take some years to materialise.

**Comment by zgrose** [comment]
FWIW, and probably a decent reason to plan on switching to .NET 8, are the End of Support dates. .NET 6 is EoS Nov 2024, .NET 7 is EoS May 2024.

**Comment by tomap** [comment]
> I appreciate the feedback and will try to canvas for some more opinions on what to do about .NET 7+.
> 
> My main concern is this. Some users of Excel-DNA create add-ins that are used beyond their organization, e.g. they ship their add-in as part of a product to outside customers, or the add-in (like the Excel-DNA IntelliSense extension) is used in various settings that might not carefully manage the .NET environment. If we support making add-ins with .NET 7+, any add-in targeting .NET 7+ will block all add-ins targeting .NET 6. And conversely, any add-in targeting .NET 6 will block all add-ins that target .NET 7+. This would mean the only viable runtime target for add-ins to be used in less controlled internal environments, or by outside users, would be .NET Framework 4.x. So, by allowing .NET 7 to be used, we remove the .NET 6 option from many scenarios. To me it would be frustrating for all those projects to be stuck on .NET Framework for many years to come.
> 
> Over the long term there is one source of hope. Microsoft has put some effort in the pre-compiled NativeAOT option, with quite a bit of progress in .NET 8. It might be possible to engineer Excel-DNA to target NativeAOT, and in that way make a native code add-in based on the .NET 8+ runtime. These would not clash with any other add-ins, freeing us from the runtime restriction. This is not trivial from the Excel-DNA side (where reflection, runtime compilation and COM features are used extensively) so might take a while. Also, a NativeAOT add-in would require all the libraries used to also work with AOT, which might be a problem for a few years, since it's all really new. So that seems like one plan, even if it will probably take some years to materialise.

Hello, 

You are talking about AOT, which would be interesting to make the standalone XLL faster, but the minimum requirement to have .Net 6 & .Net 7 & ... working together would be "self contained", not AOT, wouldn't it?

AOT or Trimming would sure improve the size of the binary and launch time, but they are not the requirement for "side by side" Excel DNA addins, aren't they?

Or myabe there is something I'm missing ?

**Comment by govert** [comment]
@tomap "the minimum requirement to have .Net 6 & .Net 7 & ... working together would be "self contained", not AOT"
No - the limitation is that there can only be one .NET core runtime loaded into the process - this is not related to whether the runtime is "self-contained" or not.
So there is no way to have both .NET 6 and .NET 7 loaded into Excel at the same time, side-by-side, in the way that .ENT Framework and .NET 6 can load side-by-side.

**Comment by lukedays** [comment]
Hi @govert! I've made some successful tests with AOT, it seems to be working well.
Check out at https://github.com/lukedays/ExcelNativeAot
Thanks in advance!!

**Comment by govert** [comment]
Hi @lukedays - yes, your project exactly what I expect for an AOT approach that avoids the version issues where possible. We (actually @Sergey-Vlasov ) have done a kind of prototype of using AOT with source generators etc. for making .NET-based SQLite extensions - the repository is here: https://github.com/SQLite-DNA/SqliteDna

For Excel-DNA I reworked the [marshaling code](https://github.com/Excel-DNA/ExcelDna/blame/master/Source/ExcelDna.Loader/XlMarshalContext.cs) a while ago to make it easy to plug into an AOT context too.

For me, the main question for AOT is how the COM integration (both ways) will work, either using the new COM source generator and / or manual COM definitions. One possible limitation with the COM support under AOT relates to the C# dynamic / IDispatch glue, which won't be supported for .NET 8, but hopefully in some future version. See my comments here: https://github.com/dotnet/runtime/issues/66674#issuecomment-1359416267

**Comment by lukedays** [comment]
Thanks for the reply! Very nice work on the SQL and marshaling codes. I'll take some inspiration from it in order to advance. 

Long shot here, what if we mimic/translate the C/C++ IDispatch implementation? I basically did this for the Excel C API calls - a rewrite of SDK methods in C#.

**Comment by govert** [comment]
For the COM / IDispatch stuff, the code I worry about is something like this
```cs
using Microsoft.Office.Interop.Excel;

class Program
{
    static void Main(string[] args)
    {
        Application app = new Application();
        app.Visible = true;

        Workbook wb = app.Workbooks.Add();
        wb.Sheets[1].Name = "FirstSheet"; 
        //        ^^^^ The code in the PIA for the indexed property wb.Sheets[...] returns "object"
        //             Under current compiler, with PIA embedding, this compiles and runs as 'dynamic'
        //             since that kicks in the dynamic /  IDispatch support.
    }
}
```

It should work fine in the .NET 2.0 style COM interop, where had to be explicit about the casts, use `get_Item` and `set_Item` etc. But that's not so pleasant. 

I don't know much about how `dynamic`, `ExpandoObject` etc. work under AOT - I have yet to experiment with this.

**Comment by lukedays** [comment]
I'll do some tests to check if the managed Interop.Excel library works. If it doesn't, maybe we can wrap the VARIANT type just like the XLOPER. 

https://learn.microsoft.com/en-US/previous-versions/office/troubleshoot/office-developer/automate-excel-from-c

**Comment by lukedays** [comment]
Interesting, I'll take a look! It's a lot of work indeed... But seems to be the way forward, since the Excel API's are being forgotten by MS (even the JS one seems abandoned).

**Comment by lukedays** [comment]
Hi @govert, do you have any tips to make ComWrappers work for Excel COM? I've searched on Github for examples but most of them seem to rely on DllImport of native objects that get wrapped afterwards.
This is the easiest example that I found, but also relies on DllImport. 
https://github.com/kant2002/CoMallocNativeAot/blob/12c1df83ec09dcd095a36fc034412cb6598b3711/CoMallocNativeAot/Program.cs
Thanks in advance!!

**Comment by govert** [comment]
I've not been brave enough to dive into the COM / AOT topic yet. I see there is the COM source generator project https://github.com/dotnet/runtime/tree/main/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator , but can't even tell how this should be used, and whether it covers both directions (exposing runtime object as COM and using COM object from the runtime). Documentation seems rather sparse.

**Comment by lukedays** [comment]
I was able to make it work for WScript.Shell, I'll now try with Excel.
https://github.com/lukedays/ComWrappersTest

**Comment by govert** [comment]
That's a great start @lukedays 

There is a bit of information about the COM source generators in the .NET 8 preview 6 release notes here https://devblogs.microsoft.com/dotnet/announcing-dotnet-8-preview-6/

I also asked a question about the source generators here: https://github.com/dotnet/runtime/issues/66674#issuecomment-1708110914

**Comment by lukedays** [comment]
Thanks @govert!
I was working exactly on that - how to use the source generator. I was able to use it to replicate the example with much less code. https://github.com/lukedays/ComWrappersTest
There are a few minor quirks:
- The interface order of method needs to be the exact same from the COM VTable.
- VARIANT needs to be custom marshalled (they seem to be working on this)
- Generic parameters on the LibraryImport calls are not supported

**Comment by lukedays** [comment]
Hi @govert, I was able to wrap `IDispatch` using a `DynamicObject`, and reproduce your example. The downside is that there's no intellisense, but a workaround might be to manually add all methods/properties from the type libraries.

https://github.com/lukedays/ExcelNativeAot/blob/master/Addin/UserFunctions.cs

Next step is to get the running Excel instance in `xlAutoOpen` to call the COM API.

**Comment by lukedays** [comment]
@govert, unfortunately `DynamicObject` doesn't work when called inside the XLL. I've thought about a few alternatives:
1) Try to regenerate the Office PIA using `GeneratedComInterface` instead of `ComImport`. The advantages would be early binding and intellisense. But properties are not supported yet and it's a lot of work (1000+ interfaces).
2) Implement the properties/methods as normal C# properties and call `IDispatch.Invoke` internally. Easier than 1 since we could cherry pick only useful props/methods but would still involve either manual find/replace or some kind of source generation. 
3) Find a workaround for `DynamicObject`.
4) Standby for now and hope for any MS solution in the future. 

Do you have any suggestions?

**Comment by govert** [comment]
@lukedays Your experiments are very interesting. Do you know why DynamicOject seemed to work from a console application, but not in an .xll? What is the problem when you try? I've not found any useful information about dynamic and NativeAOT, and have not experimented with .NET 8 yet. In any event, I don't think DynamicObject fixes (or breaks) everything - it just indicates whether we are back in the .NET 2.0 world for COM, or can try to get to the improved .NET 4 syntax etc. I'm not even sure have the details about this right.

It would be helpful if Microsoft gave some reply to my question here https://github.com/dotnet/runtime/issues/66674#issuecomment-1708110914 - that reflects my understanding and questions about the topic, and any reply would at least give some indication about their thinking. I would expect them to continue the AOT / COM work for .NET 9.

I think the first question for me is how close the new AOT-friendly COM implementation can get to supporting the important Office features - that includes properties (even including indexed properties, though that can possibly be hacked), and new-style COM events (exposing and using). It seems that the COM source generator in .NET 8 won't take us far, but I'm yet to understand whether the underlying COM plumbing is comprehensive enough if hand-writing the code.

If the only limitation is the scope of the COM source generator, I think it would be viable and useful to extend the .NET 8 COM source generator implementation to cover the whole Office use case (is this your alternative 1. ?). I see the problem as either taking the decompiled Office PIAs, or reading the PIA via reflection, and then writing out code using the ComWrappers. It would not need to be a "compile-time source generator" but could run once-off to generate the NativeAOT-interop-assembly that can be referenced as usual. That's likely to be about the same amount of work as implementing a subset of the object model by hand. Certainly, it would be more useful to work it as a source generator, and one could get other help with that since solving COM interop for Office libraries .NET 8 NativeAOT might be quite useful for many people.

**Comment by lukedays** [comment]
@govert 
> Do you know why DynamicOject seemed to work from a console application, but not in an .xll? 

I get a random error (`Unhandled exception at 0x00007FFE44AA872F (Addin.dll) in EXCEL.EXE: Unknown __fastfail() status code: 0x0000000000000048.`) that seems to be related to `System.Linq.Expressions.Interpreter`. Since only the native debugger works, it's hard to find what's the cause. Maybe something related to trimming. I'll continue trying.
[Here's an example function](https://github.com/lukedays/ExcelNativeAot/blob/9e22c881f54326febc6b39bbf1878ac8a9a49c2c/Addin/UserFunctions.cs) to test both the working method using strings and the dynamic one.

> It would be helpful if Microsoft gave some reply to my question

Yes, this would be good, so we don't reinvent the wheel if an "official" way is in the works. I'll reply to that as well.

> It would not need to be a "compile-time source generator" but could run once-off to generate the NativeAOT-interop-assembly that can be referenced as usual.

I agree. The limitations would be the get/set of COM properties, that would need to either be an extension to the `GeneratedComInterface` mechanism or indirect calls to `IDispatch.Invoke` like my example.

> Certainly, it would be more useful to work it as a source generator, and one could get other help with that since solving COM interop for Office libraries .NET 8 NativeAOT might be quite useful for many people.

For sure!!

**Comment by lilkui** [comment]
Given that .NET 8 has been released as an LTS version, I believe it is now safe to update the title of this issue to "Can you make it work with .NET 8.0?" ;D
Alternatively, could you kindly provide a tutorial on building a custom ExcelDNA version that is compatible with .NET 8?

**Comment by govert** [comment]
@lilkui I'm hoping to release a preview version that will support .NET 8 soon.

In what environment would you use it?
* Only your own machine, with no third-party add-ins that might interfere.
* Inside a company with tight control over what add-ins might run on the same machine as your add-in.
* For use outside your company, or on machines where other third-party add-ins might run.

An add-in targeting .NET 8 would break existing add-ins targeting .NET 6, or might not work if another add-in targeting .NET 6 loads first. Is this going to be a problem in your case?

**Comment by nsulikowski** [comment]
It sounds like two versions, one for .NET 8 (with the disclaimer about breaking other addins) and one for .NET 6 could be a great solution. My vote for it.

**Comment by lilkui** [comment]
> In what environment would you use it?

My environment aligns with options 1 and 2,

> An add-in targeting .NET 8 would break existing add-ins targeting .NET 6, or might not work if another add-in targeting .NET 6 loads first. Is this going to be a problem in your case?

It is not a problem for me.

> I'm hoping to release a preview version that will support .NET 8 soon.

Thank you!

**Comment by gitlsl** [comment]
@govert 

I run the hello world demo,      
` <TargetFramework>net8.0-windows</TargetFramework>`   when use .net8  , it can not work with error `                    Logger.Initialization.Error("No objects loaded from {0}", assembly.Assembly.FullName);`


just change ` <TargetFramework>net8.0-windows</TargetFramework>`  to ` <TargetFramework>net6.0-windows</TargetFramework>`    work fine

**Comment by govert** [comment]
I’ve now published a preview version of Excel-DNA to [NuGet – version 1.8.0-alpha1](https://www.nuget.org/packages/ExcelDna.AddIn/1.8.0-alpha1) – with support for projects targeting .NET 8.

For the simplest case you should be able to change the TargetFramework to 

```
    <TargetFramework>net8.0-windows</TargetFramework>

```
and the package version

```
    <PackageReference Include="ExcelDna.AddIn" Version="1.8.0-alpha1" />

```
and then everything should work as before, now targeting and running under .NET 8.

I do sometimes find build errors when changing the TargetFramework in Visual Studio, where these error go away after building again, or restarting Visual Studio.
I think VS does not always track the build tasks in NuGet packages correctly.

This version also supports the ‘RollForward’ project property for setting up some extra compatibility options.
I explain a bit more about this here: [.NET runtime support | Excel-DNA](https://excel-dna.net/docs/guides-basic/dotnet-runtime-support)
Basically it allows you to build an add-in targeting .NET 6, but allow it to load and run under .NET 8.
